python/phase-2.md

```diff
@@ -0,0 +1,831 @@
+# Python SDK - Phase 2
+
+The Python SDK will be implemented in three loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, and client
+* Phase 2 - Workflow worker, local activities, async activity completion, etc
+* Phase 3 - Deterministic workflow sandbox, replayer, test suite (some parts may be combined with phase 2)
+
+This proposal is for phase 2. It is intentionally loosely defined and all aspects are subject to discussion/change.
+
+## Overview
+
+Following up from phase 1 and known implementation changes since then, this phase will implement the actual workflow
+worker/code and other advanced features.
+
+A simple example of a workflow definition and worker run:
+
+```python
+from datetime import timedelta
+from temporalio import activity, workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+def say_hello(name: str) -> str:
+    return f"Hello, {name}!"
+
+@workflow.defn
+class MyWorkflow:
+    @workflow.run
+    async def run(arg1: str) -> str:
+        return await workflow.execute_local_activity(
```

Comment:
We can't support named parameters in workflows or activities today because our payload system has no such concept (only a collection of payloads). I already enforce this only-positional-parameter limitation today.

Reply:
But if you know the activity signature why would you want to limit calling with named params?

Reply:
Because we can't decode the param names on the other side. Similar occurs with defaults, but we let those slide as local concerns. We also want to encourage single param functions too I believe.

Someone is gonna switch the order of the params and not know that's a breaking change because they always use named params. Temporal works with positional parameters only across all SDKs, I don't think we need to rebuild the positions from the given named params.

Also, I don't want to encourage different calling styles from string name to typed. I think there's value in how similar they are today.

Reply:
I do think that it's a very reasonable expectation that we could translate kwargs to positionals given a signature but I could live with this limitation for the sake of having kwargs reserved for activity and workflow options.

Reply:
Yes, but it makes cross-SDK typing a bit weird (e.g. if you have a Go function that takes two parameters, what does that look like from a Python client that always wraps). After in-meeting discussion, I think we've settled on multiple top-level positional parameters being acceptable and serialized as such in the payload like in other SDKs.

Reply:
I don't think, at least currently, there is enough value to have payload metadata unique to the Python SDK just for named parameters. If there does become enough value, we can add it.